### PR TITLE
Removed unused USB driver parts in all usbdrv related files to shrink size by 32 bytes

### PR DESCRIPTION
**Pull request by ArminJo**
Documented bootloaderconfig.h and extended with SAVE_MCUSR which costs 6 bytes
Documented Makefile.inc and updated BOOTLOADER_ADDRESS
Added AVR_ARCHITECTURE_PARAMETER = --binary-architecture avr:35 to allow build with compiler provided by Arduino
Changed Makefile to create build.log and to use CalculateSize.py script
Replaced USB_INTR_VECTOR by USB_handler
Added Windows scripts for build convenience
Added configuration for ATtiny88
Updated all HEX files to the new and smaller version

### CODE REVIEW

File: firmware/SetPath.cmd [hunk 1]
```diff
@@ -0,0 +1,6 @@
+@set ARDUINO_DIRECTORY=E:\Elektro\arduino
```
Comment (cpldcpu) [inline firmware/SetPath.cmd:1]: This is a path that is only valid on your machine.
Why is this file needed at all?

File: firmware/make_t85_default.cmd [hunk 2]
```diff
@@ -0,0 +1,20 @@
+@echo off
```
Comment (cpldcpu) [inline firmware/make_t85_default.cmd:1]: why is this script needed? The default config can be made by "make"

File: firmware/SetPath.cmd (hunk 1 above)
Comment (ArminJo) [reply firmware/SetPath.cmd:1]: For convenience for the valued volunteer, to enable using its Arduino installation for building. AFAIK there is no default installation directory, so I choosed my own.
If you have an idea, which directory path is suited better, please tell.

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (ArminJo) [reply firmware/make_t85_default.cmd:1]: I used this script heavily for testing, since building all takes way longer until the default (reference for me) is built.

Comment (cpldcpu) [reply firmware/make_t85_default.cmd:1]: just use "make" ? It will default to t85_default.

File: firmware/SetPath.cmd (hunk 1 above)
Comment (cpldcpu) [reply firmware/SetPath.cmd:1]: hm.. I guess most people who know their way around V-USB also have a separate installation of AVR-GCC.  I think this will cause serious confusion.

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (ArminJo) [reply firmware/make_t85_default.cmd:1]: You are right!
I personally like to have a script and not to open a CMD window, navigate to the right directory and then type make.
This script is the one click solution like the old make_all.sh (it has vanished... OMG)

File: firmware/SetPath.cmd (hunk 1 above)
Comment (ArminJo) [reply firmware/SetPath.cmd:1]: So I am no people which know way around V-USB. Its OK 😀
If you find SetPath.cmd confusing, should I remove it completely?

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (cpldcpu) [reply firmware/make_t85_default.cmd:1]: You could also use your IDE to invoke build?

Comment (cpldcpu) [reply firmware/make_t85_default.cmd:1]: anyways, I think it would be better not to include environment specific scripts into the main repository. the *.cmd will fail under anything but windows.

Comment (ArminJo) [reply firmware/make_t85_default.cmd:1]: There are more than one makefile :-(

File: firmware/SetPath.cmd (hunk 1 above)
Comment (cpldcpu) [reply firmware/SetPath.cmd:1]: Er... :) Well, I have a script that opens a shell with the right path. It's basically the same, just not part of the MN repository.

I think it should be removed from the repository, because people will have different preferences how to configure their build enviroment.

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (cpldcpu) [reply firmware/make_t85_default.cmd:1]: There is only one?!?

File: firmware/Makefile [hunk 3]
```diff
@@ -7,6 +7,9 @@
 # License: GNU GPL v2 (see License.txt)
 
 CONFIG ?= t85_default
```
Comment (cpldcpu) [inline firmware/Makefile:9]: this is the only makefile and it will default to "t85_default"

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (ArminJo) [reply firmware/make_t85_default.cmd:1]: one in firmware, one in commandline?

File: firmware/SetPath.cmd (hunk 1 above)
Comment (ArminJo) [reply firmware/SetPath.cmd:1]: Done. Waiting for commit...

Comment (cpldcpu) [reply firmware/SetPath.cmd:1]: It's not added to this pull request?

File: firmware/make_t85_default.cmd (hunk 2 above)
Comment (cpldcpu) [reply firmware/make_t85_default.cmd:1]: but they are in different folders ?!?

Comment (ArminJo) [reply firmware/make_t85_default.cmd:1]: What is about the makerelease.bat script in the [2.04 release](https://github.com/micronucleus/micronucleus/releases/tag/2.04)?

There are other projects which supply these make helper scripts
Here examples from avrdude
https://github.com/mariusgreuel/avrdude/blob/windows/tools/build-mingw32.sh
and 
https://github.com/mariusgreuel/avrdude/tree/windows/windows

and Arduino
https://github.com/arduino/Arduino/blob/master/build/build_all_dist.bash

Comment (ArminJo) [reply firmware/make_t85_default.cmd]: Remove commited

File: firmware/SetPath.cmd (hunk 1 above)
Comment (ArminJo) [reply firmware/SetPath.cmd]: Remove commited

**Comment by cpldcpu** [comment]
No wait, we still need make_all. I replaced the earlier bat with that one.

**Comment by ArminJo** [comment]
> 
> 
> No wait, we still need make_all. I replaced the earlier bat with that one.

But can't you type make all or use the IDE again, like for make clean?

**Comment by Ho-Ro** [comment]
What's about Linux? Please add something like my `make_all.sh`:
````
#!/bin/sh

mkdir -p releases
mkdir -p upgrades

rm -f build.log

for config in $(ls configuration) ; do 
        make clean
        make CONFIG=${config}
        mv main.hex releases/${config}.hex
        mv upgrade.hex upgrades/upgrade-${config}.hex
done

make clean
````

**Comment by cpldcpu** [comment]
maybe we could implement this in the makefile?

**Comment by Ho-Ro** [comment]
I did a quick try, but had issues with the variable handling between (ba)sh and make, putting my code into the `Makefile` lost the `$config` content
````
every:
    for ...
        ...
    done
````
But I'm not a Makefile expert - I usually do the automation with shell scripts and not in the Makefiles.

**Comment by cpldcpu** [comment]
https://www.gnu.org/savannah-checkouts/gnu/make/manual/html_node/Foreach-Function.html

maybe this would work?

**Comment by Ho-Ro** [comment]
The Linux part works now, I just missed to use double dollar ($$config) for the variables.
````
every:
# for internal echo command, mainly on windows
ifeq ($(OS),Windows_NT)
        @echo off
        if not exist releases MKDIR releases
        if not exist upgrades MKDIR upgrades
        FOR /D  %%C IN (configuration/*) DO ( echo. \
         && echo ********************************************************** \
         && echo make Configuration %%C \
         && echo ********************************************************** \
         && make clean \
         && make CONFIG=%%C \
         && mv main.hex releases\%%C.hex \
         && mv upgrade.hex upgrades\upgrade-%%C.hex \
        )
        make clean
else
# for unix echo command
        @mkdir -p releases
        @mkdir -p upgrades
        @rm -f build.log
        @for config in `ls configuration` ; do \
          make clean; \
          make CONFIG=$$config; \
          mv main.hex releases/$$config.hex; \
          mv upgrade.hex upgrades/upgrade-$$config.hex; \
         done
        @make clean
endif
````

The Windows part is untested due to lack of a system. Not sure how to chain the commands, `;` doesn't work with `echo`.

**Comment by cpldcpu** [comment]
This seems to work well with windows, too!

I made a few smaller changes:
- removed redundant "make clean" before each build. (On windows, even rm -f will through an error sometimes when the file does not exist, so that lead to issues)
- added some "@"
-

**Comment by cpldcpu** [comment]
Ok, I pushed the new makefile to master.
https://github.com/micronucleus/micronucleus/commit/6d98d99dc0da3892d866fddf0dba053b5fa59941

The new target is name "release" -> "make release".

Thanks a lot! Finally no more platform dependent scripts. That was a long time in the making! (also, no more ruby)

**Comment by cpldcpu** [comment]
Ahh sorry... this just added more clutter instead of resolving the conflict. I am going to cherry pick manually. Sorry for that.

**Comment by ArminJo** [comment]
No problem I just replace  makefile!

**Comment by cpldcpu** [comment]
Ok, got a lot of clutter in the main repository commit history now due to all the back and forth, but everything looks fine now!

**Comment by cpldcpu** [comment]
I think we should now focus on removing the binary pushback and to also automate the hexfile release. I have no idea if it is possible to use avr-gcc with github actions. have you looked into this?

**Comment by ArminJo** [comment]
If you create a binaries repo I will see if I can change the workflow to commit it there.

**Comment by ArminJo** [comment]
Its no problem at all, at least you can use the Arduino version like I do it all the time.

**Comment by cpldcpu** [comment]
But why would we need a repo for the binary at all? The release files will always be downloadable here: 

https://github.com/micronucleus/micronucleus/releases

That's much easier than having to download a repository

**Comment by cpldcpu** [comment]
@ArminJo  I think everything is in place now. Do you think we should tag the current state as 2.5? (After updating the readme) or is still something open?

After tagging I will creat a testing branch for new pull requests.

**Comment by ArminJo** [comment]
I just test building a 32 bit windows binary, since the current one only supports 64 bit Windows versions 😞

**Comment by cpldcpu** [comment]
ah ok, then we can wait a few days. no problem.

I also found a bug in the makefile :)
